### PR TITLE
Improve stability of directional shadows when far from world origin

### PR DESCRIPTION
**Pull request by Image-unavailable**
## What problem(s) does this PR solve?

- Alleviates part of #72015
    - this isn't a perfect fix (hence "alleviates"), but it's nonetheless a significant improvement
    - this only alleviates half of the issue (poor numerical stability), the other half (FoV change issues) is addressed by [Pizzaandy](https://github.com/Pizzaandy)'s #119680 

## Additional information

This PR significantly improves the stability of directional shadows when far from the world origin.

4.5-stable:

https://github.com/user-attachments/assets/dba363a7-1207-4469-bb36-19ba64a920e6

This PR:

https://github.com/user-attachments/assets/e016a20a-ce92-41d0-8364-9af43c7235b8

### There's a Refactor in Here

This PR refactors a good chunk of the `_light_instance_setup_directional_shadow` function. I initially didn't want to do this,  but I tried many times to tweak the old code and make it work and I wasn't able to, so eventually I gave up and basically rewrote it using the old code as reference.

Part of the reason for this is that this PR (and its refactoring) is a prerequisite for another PR (#119867) which further modifies this code to improve GPU-side directional shadow performance.

As a fun side-effect of the refactoring, the code is now shorter than it was before.

### Larger Shadow Pancake

This PR introduces further padding between the shadow camera's "base" position, and the rendering position. Basically it increases shadow pancaking by a variable amount to help keep shadows more stable:

```cpp
// Snap z to very coarse steps to prevent flickering caused by z changing despite x and y being correctly pinned to a grid point.
light_view_frustum_rect_max.z = Math::snapped(cam_basis_z.dot(frustum_centroid_world) + frustum_circumscribing_radius * 2, frustum_circumscribing_radius) + pancake_size;
```

The PR would still improve the stability without this, but wouldn't be as stable due to the Z axis not being pinned at fixed intervals.

To disable variable shadow pancaking, replace the above line with the following:

```cpp
light_view_frustum_rect_max.z = cam_basis_z.dot(frustum_centroid_world) + frustum_circumscribing_radius + pancake_size;
```

Depending on whether having such large pancaking is a problem, the snap size could be reduced to smaller values if necessary (like half or quarter of the frustum radius).

### CODE REVIEW

File: servers/rendering/renderer_scene_cull.cpp [hunk 1]
```diff
@@ -2212,153 +2215,126 @@ void RendererSceneCull::_light_instance_setup_directional_shadow(int p_shadow_in
...
+		Vector3 light_view_frustum_rect_min = light_transform.basis.xform_inv(frustum_corners_local[0]);
+		Vector3 light_view_frustum_rect_max = light_transform.basis.xform_inv(frustum_corners_local[0]);
+		for (int j = 1; j < 8; j++) {
+			Vector3 cam_space_corner = light_transform.basis.xform_inv(frustum_corners_local[j]);
+			light_view_frustum_rect_min.x = MIN(light_view_frustum_rect_min.x, cam_space_corner.x);
+			light_view_frustum_rect_min.y = MIN(light_view_frustum_rect_min.y, cam_space_corner.y);
+			light_view_frustum_rect_min.z = MIN(light_view_frustum_rect_min.z, cam_space_corner.z);
+			light_view_frustum_rect_max.x = MAX(light_view_frustum_rect_max.x, cam_space_corner.x);
+			light_view_frustum_rect_max.y = MAX(light_view_frustum_rect_max.y, cam_space_corner.y);
+			light_view_frustum_rect_max.z = MAX(light_view_frustum_rect_max.z, cam_space_corner.z);
```
Comment (AThousandShips) [inline servers/rendering/renderer_scene_cull.cpp]:
```suggestion
			light_view_frustum_rect_min = light_view_frustum_rect_min.min(cam_space_corner);
			light_view_frustum_rect_max = light_view_frustum_rect_max.max(cam_space_corner);
```

**Review by Calinou — COMMENTED** [review]
Tested locally, it mostly works as expected.

When 10,000 units away from the world origin in the editor:

### Before

https://github.com/user-attachments/assets/719656e1-edd9-4b72-83d2-886e8293051e

### After

https://github.com/user-attachments/assets/43447d64-6cf5-4ed3-bfa0-96b320b29479

The variable shadow pancake can look odd when moving at the same time as a large object, as seen in the testing project and videos below.

**Testing project:** [test_shadow_stability.zip](https://github.com/user-attachments/files/29363382/test_shadow_stability.zip)

### No ground movement (for reference)

https://github.com/user-attachments/assets/26a7e741-84d7-452d-83bd-bbb8b69e534d

### With ground movement

*The camera is parented to the ground. The position displayed is the ground's origin.*

#### `master`

https://github.com/user-attachments/assets/05a7e5ff-c0d6-4064-a589-80e0ac6b0ce4

#### This PR

https://github.com/user-attachments/assets/22ca8881-9450-4bb3-8e87-158cd8d6886d

**Comment by relzz-v0** [comment]
Oh I just had this issue, glad there is a PR ready for it I hope it gets merged tho it doesn't seem that stable I haven't tested the patch tho, maybe you guys are using low qualities in your tests? It is definitely more stable tho.

**Comment by Image-unavailable** [comment]
Yeaaaaah I suspected variable pancaking could cause some issues. I hadn't thought of moving the camera alongside the scene though.

That said, I assume that people would set their pancake large enough that shadows cast properly in their scenes, even from large objects. The PR only ever makes the pancake larger or equal to the user-assigned pancake value, so scenes that didn't have issues with insufficient pancake size should keep working fine... Unless having a very large pancake causes precision issues of some kind (16-bit shadowmaps and all).

I added it in because it slightly improves shadow stability, whether it's worth it or not probably isn't for me to decide :sweat_smile:

Either way instructions for how to disable variable pancaking are in the PR post.

**Comment by Image-unavailable** [comment]
> it doesn't seem that stable

This PR works by making the calculations for shadows more precise, which makes shadows more stable when far from the origin, but there's a limit to this approach. Eventually when distance gets large enough, it stops being an issue of how precise the computation is and it becomes one of simply not having enough numbers to get the subpixel-level precision needed for stable shadows.

I'm not sure how or even if other engines deal with this kind of thing.

There might be ways to make this better or even perfect, but I can't really think of anything else, let alone getting a PR up for that :|

Do keep in mind that the shadows I showed are 8km from origin, and Calinou's are 10km. That's pretty far and you're probably going to start having other issues beyond shadows at that point.

**Comment by Image-unavailable** [comment]
Rebased and applied Skyth's changes from #120774. Turns out that with that fix applied, there was a small error on my part in how I handled PCSS, but that's now fixed.

https://github.com/user-attachments/assets/457af799-7998-473b-bb30-2b59eb5444d9

There are white bright spots appearing in shaded areas when using PCSS for some reason, but I checked master on commit 56d440ff546f1a4bf3a120d7136969f11964aea3 and confirmed the issue isn't caused by this PR.

Demo scene shown in the video:

[shadow_cull_mrp-4.8.zip](https://github.com/user-attachments/files/29816258/shadow_cull_mrp-4.8.zip)

**Comment by Image-unavailable** [comment]
Rebased.

Since I was at it, this is what Calinou's test scene looks like in this PR without variable pancaking:

https://github.com/user-attachments/assets/007bf231-17ef-41b0-8de3-913597bddde5

**Comment by BastiaanOlij** [comment]
Had a quick look at this during the rendering meeting today, code looks ok but I'd like to have some more insight into why this is better than the original. It's a bit unclear what the key point is that improved it. The location snapping might still exhibit the same behaviour but in different edge cases.

That said, it does look like testing shows improvement and I'd like to test if I can reproduce similar improvements in XR where this issue is much more noticable.

**Comment by Image-unavailable** [comment]
While I haven't done testing on this, I would say that some of the factors that contribute to the increased stability when far from the origin are:
- camera frustum size is computed in view space instead of world space, which means that its frustum is completely independent from the camera transform and will thus always have the same exact radius. This radius is then used to determine snap grid size, so if it's numerically unstable at all, like it currently is without the PR, it produces some nasty results.
- I suspect, but I am not 100% sure, that working without transform offsets and applying them only at the end helps somewhat
- there's also variable pancaking, which doesn't improve the computation accuracy by itself, but helps keep the shadow camera pinned into a fixed set of positions, which reduces flickering somewhat, though this also causes problems in some cases as Calinou's testing showed. This is easy to remove if necessary (see PR description)

Part of the reason for which the PR is the way that it is, is that's it's based on #119867 but with the draw rect code removed.

Doing all processing with `double` even in fp32 builds might give better results (I haven't tested), but that would mean not being able to use `Vector3` and `Transform3D`, which seemed like a pain in terms of code so I opted against it.

**Comment by Image-unavailable** [comment]
Squashed changes.
